### PR TITLE
Refactor code

### PR DESCRIPTION
**Pull request by ghost**
* Add consistent coding style
 * Create a new method for the Bloxberg client to send the received raw block encapsulated with our own datatype

### CODE REVIEW

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java
Comment (kiview) [inline src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:49]: Missing [Generic Type](https://docs.oracle.com/javase/tutorial/java/generics/types.html).
Type is already missing on the field though.

Note: Classes from `org.web3j` should not be exposed as part of the API if possible (hinting at the question what will be the generic type).

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java
Comment (kiview) [inline src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java:55]: Don't remove all blank lines that might have been used for prior logical grouping (e.g. it seemed sensible to have the retrieval of the iterator as part of the logical while block).

The sensible refactoring here would be to replace the iterator-while approach, this is not idiomatic for modern Java anymore.

**Review by kiview — CHANGES_REQUESTED** [review]

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java
Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:49]: The branch from the PR is a WIP one, the main idea of the commit is adding a new method in Bloxberg client to send an object of BlockWithData instead of EthBlock (from org.web3j) to whoever asks for a Block. As regards the method getTransactions, it is a WIP method just to have a working version of the code. The final version will be an object of BlockTransaction class (a custom class defined by us). However, I can change it to be generic, and for now for practical purposes send the transactions from the Ethereum block.

Comment (kiview) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:49]: If current state of the PR is a WIP and review is not called for, then please change PR to draft (there is a button for this in Github, or when creating a PR).

Also, I am not sure if you understood correctly what I said when referring to Generic Types.
We are talking about changing the return type to `List<EthBlock.Transaction>` or something alike, yes? Raw types for generic typed classes should never be used in Java.

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java
Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java:55]: It is not my priority, of course, doing so in this branch. But I need to have a coding style to have a professional code. Without it, I am using a consistent one defined by myself.

Comment (kiview) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java:55]: Just follow something along [Google Java Codestyle](https://google.github.io/styleguide/javaguide.html). 

Also just use auto format with IntelliJ default style (depends on shortcut config, but normally `Ctrl+Shift+Alt+L`,  should be fine for most of the cases https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html).

It seems not sensible to remove **all** blank lines in a method body, just want to make sure this is not an implicit coding style you are targeting. There are (and can't be) hard coding styles about when to use vertical space for organizing code logically. 

I think a good logical grouping could be the following:

```java

    public void writeOutTransactions(int blockNumber, SequenceWriter seqWriter) throws IOException {
        BigInteger blockBigInteger = BigInteger.valueOf(blockNumber);
        EthBlock.Block ethBlock = client.getEthBlock(blockBigInteger).getBlock();

        List transList = ethBlock.getTransactions();

        ListIterator it = transList.listIterator();
        while (it.hasNext()) {
            seqWriter.write(new ReducedTransObject((Transaction) it.next()));
        }
    }
```

Coding style in a team is an emergent property, therefore no other way but discuss here as part of PRs.

**Comment by kiview** [comment]
Since @lslada and I discussed this in Slack, I want to add it here, so it does not get locked from the others.

You don't want to define your own generic type here, you just use the type that you return.

Haven't run it yet, but type should be something like this, already in the field
`public List<EthBlock.TransactionResult<Transaction>> transactions;`

```
    public List<EthBlock.TransactionResult<Transaction>> getTransactions() {
        return transactions;
    }
```

(I am not sure if the generic type inside `TransactionResult` is `response.Transaction`, you need to investigate this, I just aligned with the return type of `ethBlock.getTransactions();` as documented.

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java
Comment (kiview) [inline src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:53]: This is one approach, but this is indeed potentially expensive for the user of the method (I would not expect a getter method to be this involved).

IMO doing this work already in the constructor and then having `List<BlockTransaction>` as the private field is easier and better. Or it is at least how I as and outside user of the API would expect the code to behave.

Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:53]: So, the constructor should have a the private field eg List<BlockTransaction>  block transactions? But how can someone from the outside of the constructor have access to it to make use of that list?

Comment (kiview) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/BlockWithData.java:53]: The constructor converts data from `EthBlock.Block` into our internal data structures. So the body of your current `getTransactions()` method would basically move into the constructor body (or a private method called by the constructor).

We can keep the getter method `getTransactions()`, however it should simply return the internal field `private List<BlockTransaction> transactions`, which has already been populated in the constructor.

**Comment by lslada** [comment]
Some addresses have the value null, which can already be seen in the current json file. In those cases, trying to get their code (for analyzing if they belong to a SC) gives a NullPointerException, even though I add an if-else statement. Any ideas on how I can solve this?
```java
public String getCode() throws IOException {
        String returnCode;
        if (this.address == null){
            returnCode ="nullAddress";

        }  else {
            EthGetCode code = web3j.ethGetCode(this.address, DefaultBlockParameter.valueOf("latest")).send();
             returnCode = code.getCode();
        }
        return returnCode;
    }
```

**Comment by lslada** [comment]
> Some addresses have the value null, which can already be seen in the current json file. In those cases, trying to get their code (for analyzing if they belong to a SC) gives a NullPointerException, even though I add an if-else statement. Any ideas on how I can solve this?
> 
> ```java
> public String getCode() throws IOException {
>         String returnCode;
>         if (this.address == null){
>             returnCode ="nullAddress";
> 
>         }  else {
>             EthGetCode code = web3j.ethGetCode(this.address, DefaultBlockParameter.valueOf("latest")).send();
>              returnCode = code.getCode();
>         }
>         return returnCode;
>     }
> ```
While debugging, I realized that, even though there are transactions that get the null value, the problem is that I need to use the web3j object outside the Bloxberg client. and that is what I get as null.

**Comment by kiview** [comment]
Where exactly does the NPE appear, which line in the code?
Also instead of the if-else statement, you can use `Optional`.

Not sure about the exact code, but looks something like this:

```java
Optional.of(this.address)
  .map(String a -> {
      EthGetCode code = web3j.ethGetCode(this.a, DefaultBlockParameter.valueOf("latest")).send();
      returnCode = code.getCode();
  }).orElse("")
```

**Comment by lslada** [comment]
The NPE appears in line 7 of the code I copy/pasted. The if-else statement was not the final version, it was the first one for testing purposes (make it work, make it right, make it fast). Adding a breakpoint in line 7 gives me: this.address= 0xab59a1ea1ac9af9f77518b9b4ad80942ade35088, web3j=null, which makes sense because web3j is never initialized, and is part of the Bloxberg client. The only trashy solution I can see is returning the web3j initialized object from the client, but it is something we DON'T want. Here is the full WIP of the Transaction class. I repeat that is a WIP version. 
``` java
package de.internetsicherheit.brl.bloxberg.cache.ethereum;

import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.DefaultBlockParameter;
import org.web3j.protocol.core.methods.response.EthGetCode;

import java.io.IOException;

public class TransactionAddress {
    private String address;
    private Web3j web3j;

    public TransactionAddress(String address) {
        this.address = address;
    }

    public String getAddress() {
        return address;
    }

    public String getCode() throws IOException {
        String returnCode;
        if (this.address != null) {
            EthGetCode code = web3j.ethGetCode(this.address, DefaultBlockParameter.valueOf("latest")).send();
            returnCode = code.getCode();
        } else {
            returnCode = "nullAddress";
        }
        return returnCode;
    }
}
```

**Comment by kiview** [comment]
Who do you expect this code to work, if you have web3j as an internal field and it is never assigned or created in the constructor or anywhere else?

NPE happens not because the address is null, it is because the web3j object is null.

Again this implementation has the mistake of having a side-effect and external interaction in a getter method. When creating the `TransactionAddress` object (e.g. in the `BloxbergClient` class), the fact if it has code (or even which type of address is it!) has to be determined by the assembling class that has access and knowledge about web3j.

**Comment by kiview** [comment]
Btw, although Java doesn't (currently) distinguish between plain data classes and real objects (both are implemented using `class`), you can still reflect the difference in the code and this is what I try to hint at in the previous comment.

`TransactionAddress` should be a plain data class without any logic or behaviour. It can get constructed using logic, either in a builder class that encapsulates the creation logic or having everything necessary to construct itself provided as constructor arguments. Still, there should never be things like calling out to the network (what `web3j` would do) inside a data class.

Also if it is a data class, using public fields instead of methods is, in general, okay, makes the code less verbose.

On a side note, with project Valhalla, Java will get real data classes, called records. Funnily enough, I wrote [a blog post about it at Bealdung](https://www.baeldung.com/java-valhalla-project) 😄 .

**Comment by lslada** [comment]
Yes, I mentioned that the problem was web3j variable (See the part where I mention I realized it after putting a breakpoint). 
I don't understand the part of the getter method. So, the only class that has direct access to web3j object is Bloxberg client. Should the code to know which type of Transaction be there?

**Comment by kiview** [comment]
Yes, see my previous comment. Either their or another builder like class.

**Comment by lslada** [comment]
I don't know why I need to solve conflicts of src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/ReducedTransObject.java
src/main/java/de/internetsicherheit/brl/bloxberg/cache/gui/BlockDataExtractor.java.
 The commit I am trying to PR is e1b8afbd1928ef422bfa6745dd7aac3d03e460bd (HEAD -> miner-user-sc-identification, origin/miner-user-sc-identification)

**Comment by kiview** [comment]
Because you should rebase against `master` first since #34 was merged this morning.

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/BlockDataExtractor.java
Comment (kiview) [inline src/main/java/de/internetsicherheit/brl/bloxberg/cache/BlockDataExtractor.java:26]: Only constants are written in all caps. This variable was definitely better and more idiomatic as a constant like it was in the old version of the code. 

Maybe just changing to `OUTPUT_DIRECTORY_NAME`.

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java
Comment (kiview) [inline src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java:3]: This a good example of a plain data class. Nothing to change here just wanted to highlight this.
However, I am wondering, what is the difference to `BlockTransaction `?

**Review by kiview — APPROVED** [review]
Had some comments, but would be fine to merge once merge conflicts are resolved.

File: src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java
Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java]: This was @moekappels  code for parsing the content that should be included in the JSON file. I have just commited changes to remove that class and, effectively, use BlockTransaction.

Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java]: i want the PR be made against my last commit c7bdca9a8b1ce4891cfaa95edaa6468bd1bb2a51
How can we solve this?

Comment (kiview) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java]: PR is always made between HEAD of the branch and HEAD of master. Pushing to the branch corresponding to the PR automatically updates it.

Comment (lslada) [reply src/main/java/de/internetsicherheit/brl/bloxberg/cache/ethereum/AddressInformationForJson.java]: I answer back here. After seeing @moekappels changes about the timestamp, AddressInformationForJson (which now I call it InformationForJson) is necessary because it does not include only data of the transaction.

**Comment by lslada** [comment]
The changes in the code, like using map-collector (instead of the iterator for example), have to do with performance, cleaner code?

**Comment by kiview** [comment]
Performance should be roughly the same, but this is just modern clean Java style.

**Comment by lslada** [comment]
ok, nice to know :).  So, my approach for labeling the transactions is the following. Since Bloxberg client is the one that encapsulates web3j, I would create there a new method to return the block with the transactions labelled (named getBlockWithTransactionsLabelled(block)), that will be used in writeOutTransactions() in the BlockDataAggregator.  getBlockWithTransactionsLabelled will use getBlockWithData  method inside. getBlockWithData, on the other hand, will have the transformed transactions of type BlockTransaction, each address of BlockTransaction will be of type TransactionAddress. TransactionAddress wil have two attributes (address and type), and the method to label type as SC or Other. I hope that what I wrote is clear without having coded it yet, and I would appreciate feedback :)

**Comment by kiview** [comment]
Don't make the API from `BloxbergClient` so complicated. Aim to give `BloxbergClient` maybe 2 public methods:

- `long getCurrentBlockNumber()`
- `bloxberg.cache.ethereum.Block getBlock(long blockNumber)`

Labelling should always happen when fetching a block.

**Comment by lslada** [comment]
ok, so those two methods will be the ones exposed, then the other just private to, at list, IMHO don't have a bunch of code inside a method. Do you agree with the rest of the logic?

**Comment by kiview** [comment]
You can delete the existing public methods, no need to keep them if they are not used.
Also, thinking about it, it would probably we a good idea if we also have a method `Collection<Block> getBlocks(long start, end)` or something like this, to potentially load batches.

`TransactionAddress` should not have a method to label it, it should just have an `enum` field `type`. This will be set on construction.

**Comment by lslada** [comment]
Yes, since they are used, I was just going to change the methods to private to use them internally in 
``` bloxberg.cache.ethereum.Block getBlock(long blockNumber)``` to have a cleaner code inside it. If the labelling does not happen in TransactionAddress, then, should it be made in ```transformTransactions()``` inside the class now called ```Block```?

**Comment by kiview** [comment]
Either there or in `BloxbergClient`. Just get coding, so we have concrete discussion points.

**Pull request merged**
